src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.DataFlowOperationVisitor.cs

```diff
@@ -37,52 +40,246 @@ public override GlobalFlowStateDictionaryAnalysisValue VisitLocalReference(ILoca
+                // With the initial operation as the root, expand it if
+                // the operation is parameter or local reference. It has only one pointToAnalysisResult, and it is a deferred execution invocation.
+                // Visit its argument.
+                // e.g.
+                // var a = b.Concat(c);
+                // a.ElementAt(10);
+                // When we visit 'a.Element(10)' and look back to 'b.Concat(c)', also try to visit 'b' and 'c'.
+                //
+                // Update the analysis value when reach the following nodes.
+                // 1. A parameter or local that has symbol, but no creationOperation.
+                // e.g.
+                // void Bar(IEnumerable<int> b, IEnumerable<int> c)
+                // {
+                //      var a = b.Concat(c);
+                //      a.ElementAt(10);
+                // }
+                // When 'a.ElementAt(10)' is callled, 'b' and 'c' are enumerated once.
+                // 
+                // 2. Invocation operation that returns a deferred type.
+                // e.g.
+                // void Bar()
+                // {
+                //       var a = Enumerable.Range(1, 1);
+                //       var b = Enumerable.Range(2, 2);
+                //       var c = a.Concat(b);
+                //       c.ElementAt(10);
+                // }
+                // When 'd.ElmentAt(10)' is called, think 'Enumerable.Range(1, 1)', 'Enumerable.Range(2, 2)', 'a.Concat(b)' are enumerated.
+                // 3. A parameter or local reference operation with multiple AbstractLocations. Stop expanding the tree at this node
+                // because we don't know how to proceed.
+                // e.g.
+                // void Bar(bool flag)
+                // {
+                //       var a = flag ? Enumerable.Range(1, 1) : Enumerable.Range(2, 2);
+                //       a.ElementAt(10);
+                // }
+                var queue = new Queue<IOperation>();
+                queue.Enqueue(parameterOrLocalOperation);
+                var resultAnalysisValue = defaultValue;
+                while (queue.Count > 0)
                 {
-                    return false;
+                    var currentOperation = queue.Dequeue();
+                    if (currentOperation is IParameterReferenceOperation or ILocalReferenceOperation)
+                    {
+                        var result = pointsToAnalysisResult[currentOperation];
+                        if (result.Kind != PointsToAbstractValueKind.KnownLocations || result.Locations.IsEmpty)
+                        {
+                            continue;
+                        }
+
+                        // Expand if there is only one AbstractLocation for this operation.
+                        if (result.Locations.Count == 1)
+                        {
+                            var location = result.Locations.Single();
```

Comment:
You know there is only one location so the call to `Single()` seems strange to read. I would use indexer access instead.

Reply:
This is an ImmutableHashSet, so I don't think it possible to use `result.Locations[0]` here